src/main/perf/DocGrouper.java

```diff
@@ -0,0 +1,438 @@
+package perf;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import static org.apache.lucene.tests.util.TestUtil.randomRealisticUnicodeString;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * Consuming {@link perf.LineFileDocs.LineFileDoc}, group them and put the grouped docs into
+ * a thread-safe queue.
+ */
+public abstract class DocGrouper {
+    protected final int numDocs;
+    protected final int targetNumGroups;
+    protected final BlockingQueue<DocGroups> outputQueue = new ArrayBlockingQueue<>(1024);
+    public static final DocGroups END = new DocGroups() {
+        @Override
+        public BytesRef getGroupId() {
+            return null;
+        }
+
+        @Override
+        public LineFileDocs.LineFileDoc getNextLFD() {
+            return null;
+        }
+
+        @Override
+        public int getNumOfDocsInGroup() {
+            return 0;
+        }
+
+        @Override
+        public int getRemainingNumGroups() {
+            return 0;
+        }
+    };
+
+    public static BytesRef[] group100;
+    public static BytesRef[] group100K;
+    public static BytesRef[] group10K;
+    public static BytesRef[] group1M;
+    public static BytesRef[] group2M;
+
+    DocGrouper(int numDocs, int targetNumGroups) {
+        this.numDocs = numDocs;
```

Comment:
Could we initialize the `groupIds` here instead of duplicating (differently) for the text/binary case?

And if we always just use `getTargetNumGroups` lets fix this ctor to only take `numDocs` and then call `getTargetNumGroups` (or inline it here if we are the only usage)?

Reply:
OK, let me refactor it after I removed the binary LFD from the grouper